earth_enterprise/src/common/khxml/khXML.h

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2018 the OpenGEE Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef KHXML_H
+#define KHXML_H
+
+#include <memory>
+#include "khdom.h"
+#include "notify.h"
+using namespace khxml;
+
+
+// class for enumerated types, to allow for derived classes to id themselvesx
+enum class khXMLOperationType : uint8_t
+{
+    KH_XML_READ_FILE = 0,
+    KH_XML_READ_STRING = 1,
+    KH_XML_WRITE_FILE = 2,
+    KH_XML_WRITE_STRING = 3,
+    KH_XML_LOAD_FROM_STRING = 4,
+    DEFAULT = KH_XML_READ_FILE
+};
+
+
+/*******************************************************************
+*
+*  Base class for all XML operations. Facilitates interaction with
+*  strategy contrete class.
+*
+*******************************************************************/
+class khXMLOperation
+{
+  khXMLOperation(const khXMLOperation&) = delete;
+  khXMLOperation(khXMLOperation&&) = delete;
+  khXMLOperation& operator=(const khXMLOperation&) = delete;
+  khXMLOperation& operator(khXMLOperation&&) = delete;
+
+protected:
+  std::shared_ptr<DOMDocument> doc;
+  khXMLOperationType type;
+
+public:
+  khXMLOperation(const std::string& rootTagName) throw()
+  {
+    try {
+    XMLPlatformUtils::Initialize();
+    } catch (const XMLException& e) {
+      notify(NFY_FATAL, "Unable to initialize Xerces: %s",
+             FromXMLStr(e.getMessage()).c_str());
+    }
+    try {
+
+      std::shared_ptr_ptr<DOMImplementation> impl
+              (std::move(reinterpret_cast<DOMImplementation*>
+              (DOMImplementationRegistry::getDOMImplementation(0))));
+      doc.reset(impl->createDocument(0, ToXMLStr(rootTagName), 0));
+    } catch (...) {
+          std::string error {"unable to create document"};
+          notify(NFY_FATAL,"XML error: %s", error.c_str());
+          throw khException(kh::tr(error.c_str()));
+      }
+  }
+
+  virtual ~khXMLOperation()
+  {
+    try
+    {
+      doc->release();
+    } 
+    catch (...) {}
+    if (type != khXMLOperationType::KH_XML_LOAD_FROM_STRING) khXMLOperation::Terminate();
```

Comment:
The number of Initialize calls should match the number of Terminate calls. If you Initialize unconditionally, you should Terminate unconditionally.

Reply:
Yeah, this would be taken out. I would need to get a handle on how loads are treated.